File: benchmarks/workloads/lua/mandelbrot.lua
-- benchmark: mandelbrot (LuaJIT)
-- 800x800 escape grid, escape test x^2 + y^2 >= 4, cap 255, all f64.

local function escapes(cx, cy)
  local zx, zy = 0.0, 0.0
  for i = 0, 254 do
    local zx2 = zx * zx
    local zy2 = zy * zy
    if zx2 + zy2 >= 4.0 then
      return i
    end
    local xy = zx * zy
    zy = xy + xy + cy
    zx = zx2 - zy2 + cx
  end
  return 255
end

local function workload()
  local GRID = 800
  local xmin, xmax, ymin, ymax = -2.0, 0.5, -1.25, 1.25
  local checksum = 0
  for py = 0, GRID - 1 do
    local cy = ymin + (ymax - ymin) * py / GRID
    for px = 0, GRID - 1 do
      local cx = xmin + (xmax - xmin) * px / GRID
      checksum = checksum + escapes(cx, cy)
    end
  end
  return checksum
end

-- The argv warm-up count is a minimum. Every subject also measures at least
-- 200 ms of workload execution and reports the actual warm-up time.
local WARMUP = tonumber(arg and arg[1]) or 3
local TIMED = tonumber(arg and arg[2]) or 11
local minimum_warmup = math.max(WARMUP, 3)
local warmup_seconds, warmup_iterations = 0.0, 0
local checksum = 0
while warmup_iterations < minimum_warmup or warmup_seconds < 0.200 do
  local t0 = os.clock()
  checksum = workload()
  local t1 = os.clock()
  warmup_seconds = warmup_seconds + (t1 - t0)
  warmup_iterations = warmup_iterations + 1
end
local times = {}
for i = 1, TIMED do
  local t0 = os.clock()
  checksum = workload()
  local t1 = os.clock()
  times[i] = t1 - t0
end
io.stderr:write(string.format("warmup %d %.9f\n", warmup_iterations, warmup_seconds))
for i = 1, TIMED do
  io.stderr:write(string.format("sample %d %.9f\n", i - 1, times[i]))
end
table.sort(times)
local mid = math.floor(TIMED / 2)
local median
if TIMED % 2 == 1 then
  median = times[mid + 1]
else
  median = (times[mid] + times[mid + 1]) / 2
end
io.write(string.format("%d %.9f\n", checksum, median))

File: benchmarks/workloads/c/mandelbrot.c
/* benchmark: mandelbrot (C baseline)
 * 800x800 escape-iteration grid, escape test x^2 + y^2 >= 4, cap 255, all f64.
 * Products are stored in temporaries and only added/subtracted afterwards; with
 * -ffp-contract=off no multiply-add is fused, so escape counts are bit-identical
 * across subjects. Checksum: sum of escape counts (int64).
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>

enum { GRID = 800, MAX_ITER = 255 };

static int32_t escapes(double cx, double cy) {
    double zx = 0.0, zy = 0.0;
    for (int32_t i = 0; i < MAX_ITER; i++) {
        double zx2 = zx * zx;
        double zy2 = zy * zy;
        if (zx2 + zy2 >= 4.0) {
            return i;
        }
        double xy = zx * zy;
        zy = xy + xy + cy;
        zx = zx2 - zy2 + cx;
    }
    return MAX_ITER;
}

static int64_t workload(void) {
    const double xmin = -2.0, xmax = 0.5, ymin = -1.25, ymax = 1.25;
    int64_t checksum = 0;
    for (int32_t py = 0; py < GRID; py++) {
        double cy = ymin + (ymax - ymin) * (double)py / (double)GRID;
        for (int32_t px = 0; px < GRID; px++) {
            double cx = xmin + (xmax - xmin) * (double)px / (double)GRID;
            checksum += (int64_t)escapes(cx, cy);
        }
    }
    return checksum;
}

#if defined(_WIN32)
#include <windows.h>
static double now_seconds(void) {
    LARGE_INTEGER freq, counter;
    QueryPerformanceFrequency(&freq);
    QueryPerformanceCounter(&counter);
    return (double)counter.QuadPart / (double)freq.QuadPart;
}
#else
static double now_seconds(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + (double)ts.tv_nsec * 1e-9;
}
#endif

static int cmp_double(const void *a, const void *b) {
    double x = *(const double *)a, y = *(const double *)b;
    return (x > y) - (x < y);
}

int main(int argc, char **argv) {
    /* The argv warm-up count is a minimum; measured workload execution also
     * has to reach the 200 ms methodology floor. */
    int warmup = 3, timed = 11;
    if (argc >= 3) {
        warmup = atoi(argv[1]);
        timed = atoi(argv[2]);
    }
    if (warmup < 0 || timed < 1) {
        fprintf(stderr, "usage: %s <warmup> <timed>\n", argv[0]);
        return 2;
    }
    double *times = (double *)malloc((size_t)timed * sizeof(double));
    if (times == NULL) {
        return 2;
    }
    int64_t checksum = 0;
    int warmup_iterations = 0;
    double warmup_seconds = 0.0;
    while (warmup_iterations < warmup || warmup_iterations < 3 ||
           warmup_seconds < 0.200) {
        double t0 = now_seconds();
        checksum = workload();
        double t1 = now_seconds();
        warmup_seconds += t1 - t0;
        warmup_iterations++;
    }
    /* This volatile sink makes the warm-up result observable so the optimizer
     * cannot delete the contractually required warm-up as dead code. */
    volatile int64_t warmup_sink = checksum;
    (void)warmup_sink;
    for (int i = 0; i < timed; i++) {
        double t0 = now_seconds();
        checksum = workload();
        double t1 = now_seconds();
        times[i] = t1 - t0;
    }
    fprintf(stderr, "warmup %d %.9f\n", warmup_iterations, warmup_seconds);
    for (int i = 0; i < timed; i++) {
        fprintf(stderr, "sample %d %.9f\n", i, times[i]);
    }
    qsort(times, (size_t)timed, sizeof(double), cmp_double);
    int mid = timed / 2;
    double median = (timed % 2 == 1) ? times[mid] : (times[mid - 1] + times[mid]) / 2.0;
    printf("%lld %.9f\n", (long long)checksum, median);
    free(times);
    return 0;
}
